wp-includes/mysql/class-mysql-lexer.php

```diff
@@ -0,0 +1,2752 @@
+		// considered an identifier... unless it's a float number, which ignores subsequent input.
+		$possible_identifier_prefix =
+			self::INT_NUMBER === $this->type
+			|| ( '0' === $this->text[0] && ( 'b' === $this->text[1] || 'x' === $this->text[1] ) );
+
+		if ( $possible_identifier_prefix && preg_match( '/\G' . self::PATTERN_UNQUOTED_IDENTIFIER . '/u', $this->input, $matches, 0, $start_position ) ) {
+			$end_position = $start_position + strlen( $matches[0] );
+
+			// When matched more than the number, it's an identifier.
+			if ( $end_position > $this->position ) {
+				$this->text     = $matches[0];
+				$this->position = $end_position;
+				$this->c        = $this->input[ $this->position ] ?? null;
+				$this->n        = $this->input[ $this->position + 1 ] ?? null;
+				$this->type     = self::IDENTIFIER;
+			}
+		}
+	}
+
+	protected function single_quoted_text() {
+		$pattern = $this->is_sql_mode_active( self::SQL_MODE_NO_BACKSLASH_ESCAPES )
+			? self::PATTERN_SINGLE_QUOTED_TEXT_NO_BACKSLASH_ESCAPES
+			: self::PATTERN_SINGLE_QUOTED_TEXT;
+
+		if ( preg_match( '/\G' . $pattern . '/u', $this->input, $matches, 0, $this->position ) ) {
+			$this->text      = $matches[0];
+			$this->position += strlen( $this->text );
+			$this->c         = $this->input[ $this->position ] ?? null;
+			$this->n         = $this->input[ $this->position + 1 ] ?? null;
+			$this->type      = self::SINGLE_QUOTED_TEXT;
+		} else {
+			$this->consume();
+			$this->type = self::INVALID_INPUT;
+		}
+	}
+
+	protected function double_quoted_text() {
+		$pattern = $this->is_sql_mode_active( self::SQL_MODE_NO_BACKSLASH_ESCAPES )
+			? self::PATTERN_DOUBLE_QUOTED_TEXT_NO_BACKSLASH_ESCAPES
+			: self::PATTERN_DOUBLE_QUOTED_TEXT;
+
+		if ( preg_match( '/\G' . $pattern . '/u', $this->input, $matches, 0, $this->position ) ) {
+			$this->text      = $matches[0];
+			$this->position += strlen( $this->text );
+			$this->c         = $this->input[ $this->position ] ?? null;
+			$this->n         = $this->input[ $this->position + 1 ] ?? null;
+			$this->type      = self::DOUBLE_QUOTED_TEXT;
+		} else {
+			$this->consume();
+			$this->type = self::INVALID_INPUT;
+		}
+	}
+
+	protected function back_tick_quoted_id() {
+		$pattern = $this->is_sql_mode_active( self::SQL_MODE_NO_BACKSLASH_ESCAPES )
+			? self::PATTERN_BACKTICK_QUOTED_ID_NO_BACKSLASH_ESCAPES
+			: self::PATTERN_BACKTICK_QUOTED_ID;
+
+		if ( preg_match( '/\G' . $pattern . '/u', $this->input, $matches, 0, $this->position ) ) {
+			$this->text      = $matches[0];
+			$this->position += strlen( $this->text );
+			$this->c         = $this->input[ $this->position ] ?? null;
+			$this->n         = $this->input[ $this->position + 1 ] ?? null;
+			$this->type      = self::BACK_TICK_QUOTED_ID;
+		} else {
+			$this->consume();
+			$this->type = self::INVALID_INPUT;
+		}
+	}
+
+	protected function hex_number() {
+		$is_quoted = 'x' === strtolower( $this->c ) && "'" === $this->n;
```

Comment:
`number()` already runs this check. Perhaps `hex_number()` could accept `$is_quoted` as an argument?

Reply:
Or perhaps this entire function could be inlined into `number()`. It seems like there are only two calls of it and `hex_number()` internally mimics the if/else branches already present in `number()`

Reply:
Ditto for `bin_number`